### PR TITLE
Map P90 latency and speed for "good eligible peers (>75% success rate)"

### PR DESCRIPTION
**Pull request by aarshkshah1992**

### CODE REVIEW

File: fetcher.go [hunk 1]
```diff
@@ -137,7 +151,21 @@ func (p *pool) fetchResource(ctx context.Context, from string, resource string,
 				fetchDurationPerCarPerPeerSuccessMetric.WithLabelValues(cacheStatus).Observe(float64(response_success_end.Sub(start).Milliseconds()))
 			}
 
-			updateSuccessServerTimingMetrics(respHeader.Values(servertiming.HeaderKey), resourceType, isCacheHit, durationMs, ttfbMs, received)
+			latencyMillis, speedBytesPerMs := updateSuccessServerTimingMetrics(respHeader.Values(servertiming.HeaderKey), resourceType, isCacheHit, durationMs, ttfbMs, received)
+
+			if latencyMillis > 0 && speedBytesPerMs > 0 {
+				p.lk.Lock()
```
Comment (willscott) [inline fetcher.go:159]: are we going to get in trouble if we lock full pool multiple times on every request?

File: fetcher.go [hunk 2]
```diff
@@ -295,12 +323,16 @@ func updateSuccessServerTimingMetrics(timingHeaders []string, resourceType strin
 						}
 						networkLatencyMs := ttfbMs - m.Duration.Milliseconds()
 						fetchNetworkLatencyPeerSuccessMetric.WithLabelValues(resourceType).Observe(float64(networkLatencyMs))
+						latencyMillis = float64(networkLatencyMs)
```
Comment (willscott) [inline fetcher.go:345]: can we have a separate latency tracked metric for just responses that have server cache header set?

Comment (aarshkshah1992) [reply fetcher.go:345]: @willscott Sure but should not matter here as we are measuring network latency sans the time it takes an L1 to process the request.

Comment (aarshkshah1992) [reply fetcher.go:345]: But yeah no harm in putting it.

File: metrics.go [hunk 3]
```diff
@@ -173,6 +173,18 @@ var (
 		Help:    "Network latency observed during successful caboose fetches from a single peer in milliseconds",
 		Buckets: durationMsPerCarHistogram,
 	}, []string{"resourceType"})
+
+	fetchPeerP90LatencyDistributionMetric = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:    prometheus.BuildFQName("ipfs", "caboose", "fetch_peer_network_latency_success"),
+		Help:    "Distribution of P90 network latency for a peer for network latency buckets for successful fetch from a peer",
```
[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply metrics.go:179]: done.

File: fetcher.go (hunk 1 above)
Comment (aarshkshah1992) [reply fetcher.go:159]: done.

Comment (willscott) [reply fetcher.go:159]: what are you going to use to see if it's a problem?

File: fetcher.go [hunk 4]
```diff
@@ -137,7 +153,43 @@ func (p *pool) fetchResource(ctx context.Context, from string, resource string,
...
+						fetchPeerLatencyDistributionMetric.WithLabelValues("P95").Observe(node.latencyDigest.Quantile(0.95))
+						fetchPeerLatencyDistributionMetric.WithLabelValues("P99").Observe(node.latencyDigest.Quantile(0.99))
+
+						fetchPeerSpeedDistributionMetric.WithLabelValues("P25").Observe(node.throughputDigest.Quantile(0.25))
+						fetchPeerSpeedDistributionMetric.WithLabelValues("P50").Observe(node.throughputDigest.Quantile(0.50))
+						fetchPeerSpeedDistributionMetric.WithLabelValues("P75").Observe(node.throughputDigest.Quantile(0.75))
+						fetchPeerSpeedDistributionMetric.WithLabelValues("P90").Observe(node.throughputDigest.Quantile(0.90))
+						fetchPeerSpeedDistributionMetric.WithLabelValues("P95").Observe(node.throughputDigest.Quantile(0.95))
+						fetchPeerSpeedDistributionMetric.WithLabelValues("P99").Observe(node.throughputDigest.Quantile(0.99))
+						// latency < 200ms
```
Comment (willscott) [inline fetcher.go:181]: how did we choose this magic number?

File: fetcher.go [hunk 5]
```diff
@@ -137,7 +153,43 @@ func (p *pool) fetchResource(ctx context.Context, from string, resource string,
...
+						if node.latencyDigest.Quantile(0.90) < 200 {
+							fetchPeerP90GoodLatencyCountMetric.Add(1)
+						}
+						// > 1MB/s
```
Comment (willscott) [inline fetcher.go:185]: i think as we've seen from current graphs we don't trust this one enough to feel confident in using this 'good speed count number' for anything yet

**Pull request closed without merging**